pyocd/probe/picoprobe.py

```diff
@@ -0,0 +1,636 @@
+
+    def start_queue(self):
+        # Might not need anything else.
+        self._clear_queue()
+
+
+class Picoprobe(DebugProbe):
+    """! @brief Wraps a Picolink link as a DebugProbe. """
+
+    # Address of useful DP registers.
+    SELECT = 0x8
+    RDBUFF = 0xC
+
+    # Bitmasks for AP register address fields.
+    A32 = 0x0000000c
+    APBANKSEL = 0x000000f0
+    APSEL = 0xff000000
+    APSEL_APBANKSEL = APSEL | APBANKSEL
+
+    # SWD command format
+    SWD_CMD_START = (1 << 0)    # always set
+    SWD_CMD_APnDP = (1 << 1)    # set only for AP access
+    SWD_CMD_RnW = (1 << 2)      # set only for read access
+    SWD_CMD_A32 = (3 << 3)      # bits A[3:2] of register addr
+    SWD_CMD_PARITY = (1 << 5)   # parity of APnDP|RnW|A32
+    SWD_CMD_STOP = (0 << 6)     # always clear for synch SWD
+    SWD_CMD_PARK = (1 << 7)     # driven high by host
+
+    # APnDP constants.
+    DP = 0
+    AP = 1
+
+    # Read and write constants.
+    READ = 1
+    WRITE = 0
+
+    # ACK values
+    ACK_OK = 0b001
+    ACK_WAIT = 0b010
+    ACK_FAULT = 0b100
+    ACK_ALL = ACK_FAULT | ACK_WAIT | ACK_OK
+
+    ACK_EXCEPTIONS = {
+        ACK_OK: None,
+        ACK_WAIT: exceptions.TransferTimeoutError("Picoprobe: ACK WAIT received"),
+        ACK_FAULT: exceptions.TransferFaultError("Picoprobe: ACK FAULT received"),
+        ACK_ALL: exceptions.TransferError("Picoprobe: Protocol fault"),
+    }
+
+    SAFESWD_OPTION = 'picoprobe.safeswd'
+
+    PARITY_BIT = 0x100000000
+
+    @ classmethod
+    def get_all_connected_probes(cls, unique_id=None, is_explicit=False):
+        return [cls(dev) for dev in PicoLink.enumerate_picoprobes()]
+
+    @ classmethod
+    def get_probe_with_id(cls, unique_id, is_explicit=False):
+        for dev in PicoLink.enumerate_picoprobes():
+            if dev.get_unique_id() == unique_id:
+                return cls(dev)
+
+    def __init__(self, picolink):
+        super(Picoprobe, self).__init__()
+        self._link = picolink
+        self._is_connected = False
+        self._is_open = False
+        self._unique_id = self._link.get_unique_id()
+        self._select = -1
+        self._reset = False
+
+    @ property
+    def description(self):
+        return self.vendor_name + " " + self.product_name
+
+    @ property
+    def vendor_name(self):
+        return self._link.vendor_name
+
+    @ property
+    def product_name(self):
+        return self._link.product_name
+
+    @ property
+    def supported_wire_protocols(self):
+        return [DebugProbe.Protocol.DEFAULT, DebugProbe.Protocol.SWD]
+
+    @ property
+    def unique_id(self):
+        return self._unique_id
+
+    @ property
+    def wire_protocol(self):
+        """! @brief Only valid after connecting."""
+        return DebugProbe.Protocol.SWD if self._is_connected else None
+
+    @ property
+    def is_open(self):
+        return self._is_open
+
+    @ property
+    def capabilities(self):
+        return {DebugProbe.Capability.SWJ_SEQUENCE}
```

Comment:
Let's coordinate so you can add the `SWD_SEQUENCE` capability and `swd_sequence()` API that I've added in my RP2040 branch:

https://github.com/flit/pyOCD/blob/a9b1e3824d1c809ac27c57328e943507d6cbbd99/pyocd/probe/debug_probe.py#L231

This will enable using picoprobe to debug the RP2040.

Reply:
SWD sequences:
I'll do that ASAP - probably, in less than 24h.
It seems to fit quite well with the primitives I have, so should be no great deal.
What is the max length expected?

Should I rebase the PR onto the RP2040 branch, then?
(So I get the new base class and capabilities)

As for the other points:
- SELECT: Now I understand what that capabilities meant, I was not sure there was something else to support.
I use a separate write (cached so it's probably always been skipped anyway!) to DP.SELECT, even for 'bulk' mode. I'll simply remove it. Less code, less bugs. J-link implementation, in fact, does exactly what I'm doing - well, the other way round, TBH.
- ADI: good, thanks.
- Python 2 and separate repo: If you are Ok with it, **and** support for python 2 is going to be removed, I'll just have this merged. If any of the two does not hold, I'll go for the separate repo.

Thanks for your support!

Reply:
> What is the max length expected? 

Max length is 64 cycles per sequence. There isn't a max for the number of sequences, though I need to update the CMSIS-DAP implementation to actually support this (right now it's limited to what fits in a packet, which is fine because the only user of the API is the RP2040 target for the short target selection sequence).

> Should I rebase the PR onto the RP2040 branch, then?

Yup. The other option would be to wait until I merge in the branch, then rebase on `master`.

> SELECT

Well, I messed up. 😓  The J-Link implementation doesn't actually need to deal with DP_SELECT at all. I missed removing that code when I refactored the DP_SELECT support. It still works fine, but I'm in the process of removing it right now. (This is the bug I mentioned in another comment.)

I'll add some review comments just to be thorough.

> ADIv6

I was actually sort of wrong here. The `APv2_ADDRESSES` probe capability declares support for the wider AP addresses used with ADIv6. But it only applies if the probes manages AP selection, i.e., the `MANAGED_AP_SELECTION` capability (although I declare it anyway for `CMSISDAPProbe` and soon for `JLinkProbe`, for completeness).

So once you remove the management of DP_SELECT, ADIv6 will be supported.

> Python 2

Yes, I'm happy to merge it in. It'll force me to not be lazy about removing Python 2 support! 😄

Reply:
> >  Should I rebase the PR onto the RP2040 branch, then?

> Yup. The other option would be to wait until I merge in the branch, then rebase on master.

Ok, no probs, and I've created a branch ([Picoprobe2040](https://github.com/newbrain/pyOCD/tree/PicoprobeRP2040)) from your RP2040 one, adding another remote locally to keep in synch. I've also added an implementation of swd_sequence, exercising my comprehension of list comprehension.

I'll also follow up the review here (on Picoprobe), and cherry pick the commit(s) with the review changes over PicoprobeRP2040 before making a PR there, so nothing should be lost, and you are free to point  me in whichever direction.

Once again, thanks for your great support.